### PR TITLE
implement internal safe_strcpy to replace the strncpy

### PR DESCRIPTION
**Pull request by arnopo**
Compiler complain aboiut the use of strncpy that is not safe.

Already 3 PRs propose a fix of the issue:
- openamp: replace all strncpy to strlcpy #602
- lib: fix usage of strncpy #618
- remoteproc.c: fix compile error #617

`strlcpy` seems to have been recently integrated into glibc, but it appears to only prevent overflow in the destination buffer without ensuring that we do not perform a read out of  the source buffer.

### CODE REVIEW

**Comment by arnopo** [comment]
@wyr-7, @tomi-font,
Please, Could you check that this PR fits your needs and, if so, review the PR?

File: lib/utils/string.c [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * @param src Source string to be copied.
+ * @param size Size of the destination buffer.
+ * @return The length of the source string `src`.
+ *
+ * @note If the size of the destination buffer is 0, the function does not copy any characters and
+ *       the destination buffer is not null-terminated.
+ * @note The function ensures that the destination buffer is always null-terminated if `size` is
+ *       greater than 0.
+ * @note: this code is inspired from the strlcpy.c file from freeBSD .
+ */
```
Comment (tomi-font) [inline lib/utils/string.c:32]: Why not put this in `internal/string.h`?

**Review by tomi-font — APPROVED** [review]

File: lib/utils/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+collect (PROJECT_LIB_SOURCES string.c)
```
Comment (xiaoxiang781216) [inline lib/utils/CMakeLists.txt:1]: many RTOS support strlcpy need some method to remove string.c or strlcpy from build.

Comment (arnopo) [reply lib/utils/CMakeLists.txt:1]: Do you see a potential failure in not using this RTOS strlcpy implementation? 

perhaps defining is a weak function would answer to your concern?

Comment (xiaoxiang781216) [reply lib/utils/CMakeLists.txt:1]: > Do you see a potential failure in not using this RTOS strlcpy implementation?
> 

the link will report the multiple instance error if RTOS provide one implementation.

> perhaps defining is a weak function would answer to your concern?

Yes, it's better.

Comment (arnopo) [reply lib/utils/CMakeLists.txt:1]: weak attribute added

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by edmooring — APPROVED** [review]
This looks good to go.

**Review by CV-Bowen — APPROVED** [review]

**Comment by laxiLang** [comment]
I got such AddressSanitizer error from the "string.c":

==794327==ERROR: AddressSanitizer: global-buffer-overflow on address 0x08202c63 at pc 0xed441cde bp 0xe71feba8 sp 0xe71fe778
READ of size 1 at 0x08202c63 thread T2
    #0 0xed441cdd in __interceptor_strlen ../../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:389
    #1 0x80bbf1b in strlcpy ../../../../../../../modules/lib/open-amp/open-amp/lib/utils/string.c:52
0x08202c63 is located 0 bytes to the right of global variable '*.LC24' defined in '../../../../../../../modules/lib/open-amp/open-amp/lib/rpmsg/rpmsg_virtio.c' (0x8202c60) of size 3
  '*.LC24' is ascii string 'NS'
SUMMARY: AddressSanitizer: global-buffer-overflow ../../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:389 in __interceptor_strlen


It can resolved with the changes in https://github.com/zephyrproject-rtos/open-amp/pull/23

File: lib/utils/CMakeLists.txt (hunk 2 above)
Comment (tnmysh) [reply lib/utils/CMakeLists.txt:1]: @arnopo can we introduce metal_strlcpy in libmetal and use it in open-amp instead?

That will resolve all duplicate name issues. Also, users can use default strlcpy (provided by OS or glibc).

 Please feel free to drop this idea if not suitable.

**Comment by arnopo** [comment]
> I got such AddressSanitizer error from the "string.c":
> 
> ```
> ==794327==ERROR: AddressSanitizer: global-buffer-overflow on address 0x08202c63 at pc 0xed441cde bp 0xe71feba8 sp 0xe71fe778
> READ of size 1 at 0x08202c63 thread T2
>     #0 0xed441cdd in __interceptor_strlen ../../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:389
>     #1 0x80bbf1b in strlcpy ../../../../../../../modules/lib/open-amp/open-amp/lib/utils/string.c:52
> 0x08202c63 is located 0 bytes to the right of global variable '*.LC24' defined in '../../../../../../../modules/lib/open-amp/open-amp/lib/rpmsg/rpmsg_virtio.c' (0x8202c60) of size 3
>   '*.LC24' is ascii string 'NS'
> SUMMARY: AddressSanitizer: global-buffer-overflow ../../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:389 in __interceptor_strlen
> ```
> 

@laxiLang could you tell me how to reproduce the test

> It can resolved with the changes in [zephyrproject-rtos/open-amp#23](https://github.com/zephyrproject-rtos/open-amp/pull/23)

Copy/past from comment https://github.com/zephyrproject-rtos/open-amp/pull/23/commits/cff00b649ee90c3d5cbf5e693c3b3754c736863f#r1792242968:

 The function has to return the total length of the string it tried to create so the size of the source 
 (https://man.freebsd.org/cgi/man.cgi?strlcpy(3))
 The freebsd add in code while (*src++) resulting in same overflow issue than strlen, but perhaps not...
 If you detect a potential issue, please, comment OpenAMP/open-amp#620 with your observation that I fix it in the openamp 
 repo

File: lib/utils/CMakeLists.txt (hunk 2 above)
Comment (arnopo) [reply lib/utils/CMakeLists.txt:1]: yes that seems a good alternative to address the different implementations

**Comment by laxiLang** [comment]
@arnopo : I enabled the GCC address sanitizer check. It can be easily reproduced if run a test function as below:
int main(int argc, char ** argv)
{
	union
	{
	  char src[2];
	  int  i;
	} u;
 
	u.i = 0xFFFFFFFF;
 
	char dst[100];
 
	strlcpy(dst, &u.src[0], sizeof(dst));
}
got such below error:
==904719==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fffffffda54 at pc 0x555555555299 bp 0x7fffffffd9d0 sp 0x7fffffffd9c0

READ of size 1 at 0x7fffffffda54 thread T0

    #0 0x555555555298 in strlcpy /home/jos1/repos/jos1-public/samples/strlcpy.c:11

    #1 0x555555555493 in main /home/jos1/repos/jos1-public/samples/strlcpy.c:37

    #2 0x7ffff7029d8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

    #3 0x7ffff7029e3f in __libc_start_main_impl ../csu/libc-start.c:392

    #4 0x555555555164 in _start (/home/jos1/repos/jos1-public/samples/strlcpy_sanitized+0x1164)
 
Address 0x7fffffffda54 is located in stack of thread T0 at offset 36 in frame

    #0 0x555555555397 in main /home/jos1/repos/jos1-public/samples/strlcpy.c:26
 
  This frame has 2 object(s):

    [32, 36) 'u' (line 31) <== Memory access at offset 36 overflows this variable

    [48, 148) 'dst' (line 35)

HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork

      (longjmp and C++ exceptions *are* supported)

SUMMARY: AddressSanitizer: stack-buffer-overflow /home/jos1/repos/jos1-public/samples/strlcpy.c:11 in strlcpy

Shadow bytes around the buggy address:

  0x10007fff7af0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

  0x10007fff7b00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

  0x10007fff7b10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

  0x10007fff7b20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

  0x10007fff7b30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

=>0x10007fff7b40: 00 00 00 00 00 00 f1 f1 f1 f1[04]f2 00 00 00 00

  0x10007fff7b50: 00 00 00 00 00 00 00 00 04 f3 f3 f3 f3 f3 00 00

  0x10007fff7b60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

  0x10007fff7b70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

  0x10007fff7b80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

  0x10007fff7b90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

Shadow byte legend (one shadow byte represents 8 application bytes):

  Addressable:           00

  Partially addressable: 01 02 03 04 05 06 07 

  Heap left redzone:       fa


To make address sanitizer check passed, the below change with still return total length of the string works.
 strlcpy(char *dst, const char *src, size_t size)
 {
        size_t nleft = size;
+       const char *s= src;
 
        /* Copy as many bytes as will fit. */
        if (nleft != 0) {
                while (--nleft != 0) {
-                       *dst = *src++;
+                       *dst = *s++;

**Comment by arnopo** [comment]
I have updated the PR in this way

File: lib/utils/CMakeLists.txt (hunk 2 above)
Comment (xiaoxiang781216) [reply lib/utils/CMakeLists.txt:1]: I like @tnmysh's idea, move metal_strcpy to libmetal/lib/utilities.h.

File: lib/utils/string.c [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * Copyright (c) 2024, STMicroelectronics
+ *
+ */
+
+#include <internal/string.h>
+#include <metal/io.h>
+#include <metal/utilities.h>
+
+metal_weak size_t safe_strcpy(char *dst, size_t d_size, const char *src, size_t s_size)
```
Comment (xiaoxiang781216) [inline lib/utils/string.c:12]: don't need metal_weak?

File: lib/utils/CMakeLists.txt (hunk 2 above)
Comment (arnopo) [reply lib/utils/CMakeLists.txt:1]: I am quite hesitant about that, as this new code have no OS/processor/compiler dependency.
So I would avoid to create a new libmetal API for that.
I propose to keep it in openamp ( perhaps renaming string.x to utilities.x) and move it to libmetal later, if we need different implementations

File: lib/utils/string.c (hunk 3 above)
Comment (arnopo) [reply lib/utils/string.c:12]: right i will remove it

File: lib/utils/CMakeLists.txt (hunk 2 above)
Comment (tnmysh) [reply lib/utils/CMakeLists.txt:1]: Sounds good. Yeah string.c should be renamed to utilities.c

File: lib/utils/string.c (hunk 3 above)
Comment (arnopo) [reply lib/utils/string.c]: done

File: lib/utils/CMakeLists.txt (hunk 2 above)
Comment (arnopo) [reply lib/utils/CMakeLists.txt]: done

**Review by tnmysh — APPROVED** [review]

File: lib/utils/utilities.c [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+#include <metal/utilities.h>
+
+size_t safe_strcpy(char *dst, size_t d_size, const char *src, size_t s_size)
+{
+	size_t size = metal_min(s_size, d_size);
+	size_t nleft = size + 1;
+	char *d = dst;
+
+	/* Copy as many bytes as will fit. */
+	if (nleft != 0) {
```
Comment (edmooring) [inline lib/utils/utilities.c:19]: Can this test ever be false? size_t can never be negative, so based on the assignment above, nleft must always be at least 1.

**Review by xiaoxiang781216 — APPROVED** [review]

File: lib/utils/CMakeLists.txt (hunk 2 above)
Comment (xiaoxiang781216) [reply lib/utils/CMakeLists.txt]: OK

File: lib/include/internal/utilities.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ * @param d_size Size of the destination buffer.
+ * @param src    Source string to be copied.
+ * @param s_size Size of the source buffer.
+ * @return       The length of the string contained in the `dst` buffer.
+ *
+ * @note If the size of the destination buffer is 0, the function does not copy any characters and
+ *       the destination buffer is not null-terminated.
+ * @note The function ensures that the destination buffer is always null-terminated if `size` is
+ *       greater than 0.
+ * @note The function ensures that no data is read out of the `src` buffer.
```
[the remark this answers is not in the document]
Comment (edmooring) [reply lib/include/internal/utilities.h:32]: Suggested new text:
``
@note The function ensures that no data is read past the end of the 'src' buffer.
``

[the remark this answers is not in the document]
Comment (arnopo) [reply lib/include/internal/utilities.h]: done

**Pull request merged**